onnx/version_converter/adapters/broadcast_forward_compatibility.h

```diff
@@ -22,14 +22,23 @@ class BroadcastForwardCompatibility final : public Adapter {
         const std::vector<Dimension>& B_sizes = inputs[1]->sizes();
         // Also assert that broadcasting syntax are correct if axis is not present
         if (node->hasAttribute(kaxis)) {
-          if (node->i(kaxis) != (int) (A_sizes.size() - B_sizes.size())) {
+          int axis = node->i(kaxis);
+          if (axis != (int) (A_sizes.size() - B_sizes.size())) {
             // Add a Reshape node before input B
             Node * n = graph->create(kUnsqueeze);
             n->addInput(inputs[1]);
             std::vector<int64_t> axes;
-            std::vector<Dimension> new_sizes = B_sizes;
-            for (int i = 0; i < (int) (A_sizes.size() - B_sizes.size()); i++) {
-              axes.emplace_back(B_sizes.size() + i);
+            std::vector<Dimension> new_sizes = {};
+            new_sizes.reserve(A_sizes.size());
+            // Add dimensions to the left
+            for (int i = 0; i < axis; i++) {
+              axes.emplace_back(i);
+              new_sizes.emplace_back(Dimension(1));              
+            }
+            new_sizes.insert(new_sizes.end(), B_sizes.begin(), B_sizes.end());
+            // Add dimensions to the right
+            for (int i = 0; i < (int) (A_sizes.size() - B_sizes.size() - axis); i++) {
+              axes.emplace_back(B_sizes.size() + axis + i);
               new_sizes.emplace_back(Dimension(1));
```

Comment:
I went looking for the old broadcast behaviour [here](https://github.com/onnx/onnx/blob/master/docs/Changelog.md#Add-6), which states that dimensions could also be added to the left of the shape list. But the previous adapter was only adding them to the right.

Reply:
I guess the reason is the old broadcast behaviour (opset 6) only supports limited broadcast (only adding dimensions to the right one). 

If the op is upgraded to opset 7, it will allow Multi-directional Broadcasting. Therefore, the version_converter needs to do Unidirectional Broadcast (only adding dimensions to the right one) for opset 6 first. Otherwise, opset 7 would misunderstand how to broadcast these old ops. 

IMO, adding dimensions to the left here looks unnecessary to me.

Reply:
By "left" and "right" I don't mean the left and right inputs of the node, but left and right of the dimension list of the right (second) input, i.e. prepending (left) and appending (right) dimensions. What I added is still adding dimensions to B only, but allows adding them on both sides of the dimension list. Previously, this adapter was only adding them to the right (i.e. appending them, via `axes.emplace_back(B_sizes.size() + i)`).

The examples in the changelog I linked are both prepending and appending dimensions:
```
shape(A) = (2, 3, 4, 5), shape(B) = (,), i.e. B is a scalar tensor
shape(A) = (2, 3, 4, 5), shape(B) = (1, 1), i.e. B is an 1-element tensor
shape(A) = (2, 3, 4, 5), shape(B) = (5,)
shape(A) = (2, 3, 4, 5), shape(B) = (4, 5)
shape(A) = (2, 3, 4, 5), shape(B) = (3, 4), with axis=1
shape(A) = (2, 3, 4, 5), shape(B) = (2), with axis=0
```

Reply:
Ah I realize that the reason why dimensions were not prepended might be that opset 7 (and higher versions) allow prepending dimensions themselves, so there's no reason to explicitly prepend them in the adapter. Is that correct? I will revert this change if yes.

Reply:
Sorry for my previous misunderstanding.
But I think the answer is yes. The key point is for broadcast in opset 7, it does not need an axis to append anymore.
Besides, as you mentioned, it can handle prepending dimensions. 
So the only thing we need to do here for opset 6 to 7 is appending the dimensions by the axis.

Reply:
Reverted